onnxruntime/core/providers/cpu/tensor/expand.cc

```diff
@@ -0,0 +1,157 @@
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<TYPE>()), \
+      Expand<TYPE>);                                                               \
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                                  \
+      Expand,                                                                      \
+      13,                                                                          \
+      TYPE,                                                                        \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<TYPE>()), \
+      Expand<TYPE>);
+
+REG_EXPAND_KERNEL(float)
+REG_EXPAND_KERNEL(double)
+REG_EXPAND_KERNEL(int8_t)
+REG_EXPAND_KERNEL(int16_t)
+REG_EXPAND_KERNEL(int32_t)
+REG_EXPAND_KERNEL(int64_t)
+REG_EXPAND_KERNEL(uint8_t)
+REG_EXPAND_KERNEL(uint16_t)
+REG_EXPAND_KERNEL(uint32_t)
+REG_EXPAND_KERNEL(uint64_t)
+REG_EXPAND_KERNEL(bool)
+REG_EXPAND_KERNEL(MLFloat16)
+
+template <typename T>
+Status Expand<T>::Compute(OpKernelContext* context) const {
+  const auto* input_tensor = context->Input<Tensor>(0);
+  const auto* input_data = input_tensor->template Data<T>();
+  const auto& input_shape = input_tensor->Shape().GetDims();
+
+  const auto* input_dims = input_shape.data();
+  auto input_dims_size = static_cast<int64_t>(input_shape.size());
+
+  const auto* shape_tensor = context->Input<Tensor>(1);
+  const auto* shape_dims = shape_tensor->Data<int64_t>();
+  std::vector<int64_t> output_shape{shape_dims, shape_dims + shape_tensor->Shape().Size()};
+
+  if (input_shape.size() > output_shape.size()) {
+    output_shape.insert(output_shape.begin(), input_shape.size() - output_shape.size(), 1);
+  }
+
+  auto input_shape_iter = input_shape.rbegin();
+  auto output_shape_iter = output_shape.rbegin();
+  while (input_shape_iter != input_shape.rend() &&
+         output_shape_iter != output_shape.rend()) {
+    if (*input_shape_iter != *output_shape_iter) {
+      if (1 == *output_shape_iter) {
+        *output_shape_iter = *input_shape_iter;
+      } else if (1 != *input_shape_iter) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "invalid expand shape");
+      }
+    }
+    input_shape_iter++;
+    output_shape_iter++;
+  }
+
+  TensorShape output_tensor_shape(output_shape);
+  auto* output_tensor = context->Output(0, output_tensor_shape);
+  auto* output_data = output_tensor->template MutableData<T>();
+  auto* output_dims = output_shape.data();
+  auto output_dims_size = static_cast<int64_t>(output_shape.size());
+  auto max_dims_size = std::max(input_dims_size, output_dims_size);
+  std::unique_ptr<int64_t[]> input_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> output_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> expand_dim_size{new int64_t[max_dims_size]};
+  auto dim_group_start = max_dims_size;
+
+  for (int64_t input_dims_iter = input_dims_size - 1,
```

Comment:
Put all these initalizers before this for loop?

Reply:
It is reasonable suggestion indeed while they are purely looping variables ...